Zipkin library

fabolive opened the pull request:
This PR adds a Zipkin library to the Envoy code base. Changes made to Envoy to use this library to support Zipkin tracing are in PR #700. This needs to be merged before PR #700.

cc: @RomanDzhabarov, @adriancole, @mattklein123

mattklein123 commented:
💥 very excited for this! Will let @RomanDzhabarov take first pass.

rshriram commented:
@htuch @rlazarus there is a source/zipkin folder. The code there is supposed to be a generic library. I am not sure whats the best way to bazel it up.

RomanDzhabarov commented:
@fabolive this is very exciting. Is there any chance to split this into multiple self-contained PRs?

rshriram commented:
@RomanDzhabarov one thing that we were discussing: the source/zipkin folder is a separate library. That could go in as its own PR. The rest is additions to http tracer. How does that sound?

rshriram commented:
cross referencing #628

htuch commented:
I'd follow other examples of using `envoy_cc_library` and writing `BUILD` rules to Bazelify, see https://github.com/lyft/envoy/blob/master/bazel/DEVELOPER.md

RomanDzhabarov commented:
Yup, that high level separation sounds good.

fabolive commented:
@RomanDzhabarov I will then re-purpose this branch and create another one on top.

rshriram commented:
Just FYI, here is the front-proxy example with zipkin integration. https://gist.github.com/rshriram/15f05f530894eff3cebf527ca1ba5857

adriancole commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -453,6 +453,7 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
                                               connection_manager_.config_.tracingStats());
 
     if (tracing_decision.is_tracing) {
+      stream_log_debug("XXX Fabio: About to call startSpan", *this);
```
prefer two XX

..
just kidding :P

adriancole commented on source/common/tracing/http_tracer_impl.cc:
```diff
@@ -320,4 +324,194 @@ void LightStepRecorder::onSuccess(Http::MessagePtr&& msg) {
+    std::cerr << std::endl
+              << "Found context" << std::endl;
+
+    // Get the context from the x-b3-envoy header
```
is this comment out-of-date? I don't see x-b3-envoy listed in the code.

fabolive replied:
It is out of date, indeed.

adriancole commented on source/zipkin/zipkin_core_types.cc:
```diff
@@ -0,0 +1,255 @@
+    // Need to set the SS annotation
+    Annotation ss;
+    ss.setHost(annotations_[0].host());
+    ss.setTimestamp(Util::timeSinceEpochMicro());
```
timestamp and duration should only be set on server spans which are not client-originated (ex root spans which weren't resumed from headers)

https://github.com/openzipkin/openzipkin.github.io/issues/49

In brave, we set a flag in the in-process struct representing the context called "shared" which is used to differentiate a client-originated root span from a server-originated one. (the former is shared, the latter is not). When a trace context is parsed from headers, shared is set to true. If it wasn't it defaults to false.

```java
  /**
   * True if we are contributing to a span started by another tracer (ex on a different host).
   * Defaults to false.
   *
   * <p>When an RPC trace is client-originated, it will be sampled and the same span ID is used for
   * the server side. However, the server shouldn't set span.timestamp or duration since it didn't
   * start the span.
   */
  public abstract boolean shared();
```

We use this flag when creating the span (like it is here)
```
    // don't report server-side timestamp on shared or incomplete spans
    if (shared && (flags & FLAG_SR) != 0) {
      span.timestamp(null).duration(null);
    }
```

fabolive replied:
@adriancole The code you highlighted above is setting the timestamp of the annotation, not the span's. I make sure that only the span initiator sets the timestamp and duration of a span, as explained here: http://zipkin.io/pages/instrumenting.html

adriancole replied:
lol yes.. you are right! eyes blurred

adriancole commented on source/zipkin/zipkin_core_types.h:
```diff
@@ -0,0 +1,284 @@
+
+  const Endpoint& host() const { return host_; }
+
+  void setHost(const Endpoint& host) {
```
endpoint is the proper name for this field (we haven't renamed the field in thrift, but all docs refer to it as endpoint iirc)

fabolive replied:
Fair enough.

adriancole commented on source/zipkin/zipkin_core_types.h:
```diff
@@ -0,0 +1,284 @@
+  bool isset_host_;
+};
+
+enum AnnotationType { BOOL = 0, BYTES = 1, I16 = 2, I32 = 3, I64 = 4, DOUBLE = 5, STRING = 6 };
```
you can simplify this by only supporting BOOL and STRING as others are not used in practice in zipkin.

BOOL is a special case for identifying the peer address. for example, "ca", true, host=client address, not the local address.

https://github.com/openzipkin/zipkin-api/blob/master/thrift/zipkinCore.thrift#L226

fabolive replied:
OK.

adriancole commented on source/zipkin/zipkin_core_types.cc:
```diff
@@ -0,0 +1,255 @@
+  if ((context.isSetAnnotation().sr) && (!context.isSetAnnotation().ss)) {
+    // Need to set the SS annotation
+    Annotation ss;
+    ss.setHost(annotations_[0].host());
```
on all the core annotations "cs" "sr" "ss" "cr" as well normal binary annotations (string value), the host will be constant for the tracer, as it is identifying the local process. I can't see clearly if that's the case in the current variant of this code, so just sayin..

There are special annotations "ca" and "sa" which identify the peer (incoming client or downstream server), which are optional but useful if you know it. I made a comment elsewhere on that.

fabolive replied:
The current code guarantees that the same endpoint (identifying the envoy in question) is used for the annotations originating from the same envoy process.
Indeed, ca/sa might be useful.

fabolive commented:
@RomanDzhabarov We have 2 PRs now. This one has been re-purposed to contain only the addition of a Zipkin library to the Envoy code base. PR #700, in turn, contains the changes to make Envoy use this library.

cc: @adriancole, @rshriram, @mattklein123

htuch commented on source/zipkin/BUILD:
```diff
@@ -0,0 +1,28 @@
+        "zipkin_json_field_names.h",
+    ],
+    external_deps = ["rapidjson"],
+    deps = [],
```
You can just skip the `deps` if empty.

htuch commented on source/zipkin/BUILD:
```diff
@@ -0,0 +1,28 @@
+        "span_context.h",
+        "tracer.h",
+        "tracer_interface.h",
+        "util.h",
```
If any of these are purely internal to the library and don't need exporting for the use of other libraries, you can move them into `srcs`.

adriancole commented on source/zipkin/zipkin_core_types.cc:
```diff
@@ -0,0 +1,255 @@
+  return json_string_;
+}
+
+void Span::finish() {
```
when an operation abends, is finish still called? if so, we'd usually put a string binary annotation of key "error" and value of the error message in the span

fabolive replied:
In Envoy, when a connection is recycled the connection manager checks for errors. If an error is caught, it calls span.setTag(error, true) to indicate that an error has occurred. In my code, span.setTag(key, value) essentially adds a binary annotation to the span. So, the code behaves just like you were expecting. 
setTag: https://github.com/amalgam8/envoy/blob/zipkin/source/zipkin/zipkin_core_types.cc#L231

adriancole commented on test/zipkin/span_context_test.cc:
```diff
@@ -0,0 +1,152 @@
+  EXPECT_FALSE(spanContext.isSetAnnotation().ss);
+  EXPECT_EQ("0000000000000000;0000000000000000;0000000000000000", spanContext.serializeToString());
+
+  // Span context populated with trace id, id, parent id, and no annotations
```
not sure if here or not, but should test that 128bit variant doesn't crash. All new instrumentation support 128-bit trace ids, so if you accept an inbound trace id, the choices are to truncate by dropping high bits, or better.. use all of them.

fabolive replied:
I have added a test to verify this  behavior. Note that the Span class has a member named traceIdHigh_, plus a flag indicating whether or not it has been set. That's how a Span with a 128-bit trace id can be created. Currently, the code ignores the higher bits.
Here is the  test I added:
https://github.com/amalgam8/envoy/blob/zipkin/test/zipkin/span_context_test.cc#L117

adriancole commented on test/zipkin/tracer_test.cc:
```diff
@@ -0,0 +1,168 @@
+  EXPECT_NE(0ULL, ann.timestamp()); // annotation's timestamp must be set
+  EXPECT_TRUE(ann.isSetEndpoint());
+  Endpoint endpoint = ann.endpoint();
+  EXPECT_EQ("127.0.0.1", endpoint.ipv4());
```
presumably for json encoding bc thrift it is packed into a 32bit unsigned

fabolive replied:
Correct. Since the current implementation supports only JSON over HTTP, we are storing IP addresses as strings. When we support other transports, we will revisit that (and possibly a few other aspects of the core data structures).

adriancole commented on test/zipkin/tracer_test.cc:
```diff
@@ -0,0 +1,168 @@
+  EXPECT_FALSE(server_side_shared_context_span.isSet().duration);
+
+  // ==============
+  // Test the creation of a child span --> CS
```
this assumes the only sort of child span is a client span. That's ok if that's the intent

fabolive replied:
You are spot on! The other scenario is missing from the test case; I was aware of it and was planning to add it today.  :-)

rshriram commented on source/zipkin/tracer.cc:
```diff
@@ -103,7 +99,7 @@ Span Tracer::startSpan(const std::string& span_name, uint64_t start_time,
 
   // Add the newly-created annotation to the span
-  annotation.setHost(std::move(ep));
+  annotation.setEndpoint(std::move(ep));
```
I suggest using full names instead of shorthands. E.g. endpoint instead of ep. Annotation instead of ann.

rshriram commented on source/zipkin/tracer.cc:
```diff
@@ -0,0 +1,137 @@
+  span.setStartTime(start_time);
+
+  // Set the timestamp globally for the span and also for the CS annotation
+  timestampMicro = Util::timeSinceEpochMicro();
```
Micro -> US or if it's not too lengthy, might as well use Microseconds

rshriram commented on source/zipkin/zipkin_core_types.cc:
```diff
@@ -0,0 +1,236 @@
+  trace_id_high_ = span.traceIdHigh();
+  isset_ = span.isSet();
+  start_time_ = span.startTime();
+  tracer_ = span.tracer();
```
Do we need this copy constructor if it's just a one to one assignment of ints and pointers to const objects ?

fabolive commented:
@adriancole: Have I answered your questions to your satisfaction? Have I addressed properly your concerns?
@RomanDzhabarov This is no longer a work-in-progress PR. I consider it ready, unless other issues are raised or the reviewers are not satisfied with my answers and latest commits that attempt to address their concerns.

cc: @rshriram

mattklein123 commented on source/CMakeLists.txt:
```diff
@@ -1,3 +1,4 @@
 add_subdirectory(common)
 add_subdirectory(exe)
 add_subdirectory(server)
+add_subdirectory(zipkin)
```
I don't think I would add this as a top level directory. Please move to source/common/tracing/zipkin. There are other things we may eventually want to compile out via various build options that are already in common. We can deal with that at that point.

mattklein123 commented on source/zipkin/span_buffer.cc:
```diff
@@ -0,0 +1,40 @@
+    // Buffer full
+    return false;
+  }
+  // span_buffer_[next_position_++] = &span;
```
delete?

mattklein123 commented on source/zipkin/span_buffer.cc:
```diff
@@ -0,0 +1,40 @@
+}
+
+std::string SpanBuffer::toStringifiedJsonArray() {
+  std::string stringifiedJsonArray = "[";
```
All variables should use snake case. I won't comment on this further but please fix everywhere.

mattklein123 commented on source/zipkin/span_buffer.h:
```diff
@@ -0,0 +1,36 @@
+#pragma once
+
+#include <vector>
```
Remove all standard headers that are already in precompiled.h (I won't comment on this in other places).

mattklein123 commented on source/zipkin/span_buffer.h:
```diff
@@ -0,0 +1,36 @@
+
+namespace Zipkin {
+
+typedef std::shared_ptr<Span> SpanPtr;
```
SpanSharedPtr (there might be other instances of this, won't comment further).

mattklein123 commented on source/zipkin/zipkin_json_field_names.cc:
```diff
@@ -0,0 +1,23 @@
+#include "zipkin/zipkin_json_field_names.h"
+
+const std::string Zipkin::ZipkinJsonFieldNames::SPAN_TRACE_ID = "traceId";
```
nit: put inside Zipkin namespace to avoid extra typing.

mattklein123 commented on source/zipkin/span_context.h:
```diff
@@ -0,0 +1,53 @@
+  bool sr : 1;
+} Annotation_isset_;
+
+class SpanContext {
```
To make the code easier to read/review, please have at minimum class level comments for each class/struct. Optimally, every public function should have proper doc comments to make it easier to review. (This applies everywhere, won't comment futher).

mattklein123 commented on source/zipkin/tracer.cc:
```diff
@@ -0,0 +1,137 @@
+  Endpoint ep;
+  uint64_t timestampMicro;
+
+  // TODO We currently ignore the start_time to set the span/annotation timestamps
```
Please use TODO(your GH name): style for TODOs. (Applies elsewhere)

mattklein123 commented on source/zipkin/tracer.cc:
```diff
@@ -0,0 +1,137 @@
+  } else {
+    // Unexpected condition
+
+    // TODO Log an error
```
You should have a stat here. If there are similar TODOs please also make sure there are error stats.

mattklein123 commented on source/zipkin/util.cc:
```diff
@@ -0,0 +1,66 @@
+  return rand_64();
+}
+
+std::string Util::uint64ToBase16(uint64_t value) {
```
This is very slow, use common/hex.h instead.

adriancole commented:
>
> @adriancole <https://github.com/adriancole>: Have I answered your
> questions to your satisfaction? Have I addressed properly your concerns?
>
Yep! should this change include a readme update on how to test the code
locally? or any other description of this feature?

RomanDzhabarov commented on source/zipkin/span_buffer.h:
```diff
@@ -0,0 +1,36 @@
+
+  SpanBuffer(uint64_t size) : span_buffer_(size, nullptr), size_(size), next_position_(0) {}
+
+  virtual ~SpanBuffer() {}
```
need comments on the methods

RomanDzhabarov commented on source/zipkin/span_context.cc:
```diff
@@ -0,0 +1,107 @@
+  id_ = span.id();
+  parent_id_ = span.parentId();
+
+  for (Annotation ann : span.annotations()) {
```
full name for 'ann'

RomanDzhabarov replied:
also should not that be, Annotation& annotation

RomanDzhabarov commented on source/zipkin/span_context.cc:
```diff
@@ -0,0 +1,107 @@
+    return "0000000000000000;0000000000000000;0000000000000000";
+  }
+
+  std::string s;
```
s/s/result or something

RomanDzhabarov commented on source/zipkin/span_context.cc:
```diff
@@ -0,0 +1,107 @@
+  return s;
+}
+
+void SpanContext::populateFromString(std::string s) {
```
const std::string&

RomanDzhabarov commented on source/zipkin/span_context.h:
```diff
@@ -0,0 +1,53 @@
+
+typedef struct Annotation_isset_ {
+  Annotation_isset_() : cs(false), cr(false), ss(false), sr(false) {}
+  bool cs : 1;
```
struct members should have _ at the end

RomanDzhabarov commented on source/zipkin/tracer.cc:
```diff
@@ -0,0 +1,137 @@
+  reporter_ = std::shared_ptr<Reporter>(std::move(reporter));
+}
+
+void Tracer::getIPAndPort(const std::string& address, std::string& ip, uint16_t& port) {
```
this feels like some common utility method

RomanDzhabarov commented on source/zipkin/tracer.h:
```diff
@@ -0,0 +1,56 @@
+  /**
+   * Span-buffer management policy to be implemented by users of the Trace class
+   */
+  virtual void reportSpan(Span&& span) = 0;
```
PURE is the #define we use here for pure virtual methods

fabolive commented:
@mattklein123, @RomanDzhabarov: I believe I have addressed all your comments above. I have added comments to all classes, structs, enums, and public methods. I  have also addressed your requests for optimizations, stylistic changes, and code restructuring.
cc: @rshriram

mattklein123 commented:
@fabolive thanks I will go through this in detail tomorrow. @RomanDzhabarov please do so as well.

RomanDzhabarov commented on source/common/tracing/zipkin/span_buffer.h:
```diff
@@ -0,0 +1,72 @@
+   * Empties the buffer. This method is supposed to be called when all buffered spans
+   * have been sent to to the Zipkin service.
+   */
+  void flush();
```
my be call this clear() for expressing intent better

fabolive replied:
Done.

RomanDzhabarov commented on source/common/tracing/zipkin/span_buffer.h:
```diff
@@ -0,0 +1,72 @@
+private:
+  // We use a pre-allocated vector to improve performance
+  std::vector<SpanSharedPtr> span_buffer_;
+  uint64_t size_;
```
why do we have explicit size_, it should be available as span_buffer_.size()? (this seems redundant)

fabolive replied:
Fixed.

RomanDzhabarov commented on source/common/tracing/zipkin/span_context.cc:
```diff
@@ -0,0 +1,103 @@
+    }
+  }
+
+  is_populated_ = true;
```
nit, is_initialized

fabolive replied:
Done

RomanDzhabarov commented on source/common/tracing/zipkin/span_context.cc:
```diff
@@ -0,0 +1,103 @@
+
+const std::string SpanContext::serializeToString() {
+  if (!is_populated_) {
+    return "0000000000000000;0000000000000000;0000000000000000";
```
can you create static const and name it "UNINITIALIZED_CONTEXT" or something along this line

fabolive replied:
Done

RomanDzhabarov commented on source/common/tracing/zipkin/span_context.cc:
```diff
@@ -0,0 +1,103 @@
+void SpanContext::populateFromString(const std::string& s) {
+  // ^([0-9,a-z]{16});([0-9,a-z]{16});([0-9,a-z]{16})((;(cs|sr|cr|ss))*)$
+  std::string hex_digit_group = "([0-9,a-z]{16})";
+  std::string reg_ex_str =
```
make this static

fabolive replied:
Done

RomanDzhabarov commented on source/common/tracing/zipkin/span_context.cc:
```diff
@@ -0,0 +1,103 @@
+      ZipkinCoreConstants::SERVER_RECV + "|" + ZipkinCoreConstants::CLIENT_RECV + "|" +
+      ZipkinCoreConstants::SERVER_SEND + "))*)$";
+
+  std::regex re(reg_ex_str);
```
ditto

fabolive replied:
Done

RomanDzhabarov commented on source/common/tracing/zipkin/span_context.cc:
```diff
@@ -0,0 +1,103 @@
+    std::string matched_annotations = match.str(4);
+    if (matched_annotations.size() > 0) {
+      char* annotation_value_strings = const_cast<char*>(matched_annotations.c_str());
+      char* annotation_value = std::strtok(annotation_value_strings, FIELD_SEPARATOR_.c_str());
```
why do you need this? can't you just use StringUtils::Split?

fabolive replied:
Using StringUtil::split now

RomanDzhabarov commented on source/common/tracing/zipkin/span_context.h:
```diff
@@ -0,0 +1,123 @@
+ * SS: "Server Send"
+ * SR: "Server Receive"
+ */
+typedef struct Annotation_isset_ {
```
struct name has incorrect style

fabolive replied:
Fixed

RomanDzhabarov commented on source/common/tracing/zipkin/span_context.h:
```diff
@@ -0,0 +1,123 @@
+  /**
+   * Constructor that creates a context object from the given Zipkin span object.
+   *
+   * @param span The Zipkin span used to initialize a SpanContext object
```
nit: '.' at the end everywhere

fabolive replied:
Fixed

RomanDzhabarov commented on source/common/tracing/zipkin/tracer.cc:
```diff
@@ -0,0 +1,122 @@
+namespace Zipkin {
+
+Span Tracer::startSpan(const std::string& span_name, uint64_t start_time) {
+  Span span;
```
in general we have declaration close to the usage, can you move those around?

fabolive replied:
Done

RomanDzhabarov commented on source/common/tracing/zipkin/util.cc:
```diff
@@ -0,0 +1,84 @@
+             std::chrono::system_clock::now().time_since_epoch()).count();
+}
+
+uint64_t Util::generateRandom64() {
```
we already have RNG, can you reuse that?

fabolive replied:
I have modified the code to optionally take a custom random-number generated. The code in PR #700 will make sure to call tracer.setRandomGenerator and pass the random-generator currently used by Envoy.

RomanDzhabarov commented on source/common/tracing/zipkin/util.cc:
```diff
@@ -0,0 +1,84 @@
+}
+
+void Util::getIPAndPort(const std::string& address, std::string& ip, uint16_t& port) {
+  std::regex re("^(\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3})(:(\\d+))?$");
```
we also should support ipv6 i think, but i'm not sure how far we are in ipv6 support @mattklein123

fabolive replied:
I suggest that we revisit IPv6 support later on.

mattklein123 replied:
That's fine.

mattklein123 replied:
Please add TODO() that we need to add IPv6 support. cc @jamessynge

RomanDzhabarov commented on source/common/tracing/zipkin/util.h:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Utility class with a few convenient methods
+ */
+class Util {
```
this class looks like a sink for all possible utility methods :-)
I think some of those methods are already in existing code base, can you reuse (i point to some examples above)

RomanDzhabarov commented on source/common/tracing/zipkin/util.h:
```diff
@@ -0,0 +1,73 @@
+   *
+   * @param value The integer to be converted
+   */
+  static std::string uint64ToBase16(uint64_t value);
```
uint64ToHex may be?

RomanDzhabarov commented on source/common/tracing/zipkin/zipkin_core_constants.cc:
```diff
@@ -0,0 +1,35 @@
+const std::string ZipkinCoreConstants::SERVER_RECV = "sr";
+const std::string ZipkinCoreConstants::WIRE_SEND = "ws";
+const std::string ZipkinCoreConstants::WIRE_RECV = "wr";
+const std::string ZipkinCoreConstants::CLIENT_SEND_FRAGMENT = "csf";
```
where is this going to be used?

adriancole replied:
might be good to lazy declare these, since many don't use them at all and could be distracting. "ws" "wr" is only used in finagle on the client side (when the bytes hit the wire as opposed to when the client starts creating the request).. this feature could be punted unless you have a good hook for this.

fabolive replied:
@RomanDzhabarov do you prefer the constants unlikely to be used by Envoy be removed?

mattklein123 replied:
Yes please delete any constants/code that is not used.

RomanDzhabarov commented on test/common/tracing/zipkin/span_buffer_test.cc:
```diff
@@ -0,0 +1,100 @@
+
+  buffer.addSpan(std::move(span));
+  EXPECT_EQ(1ULL, buffer.pendingSpans());
+  std::string expected_json_array_string = ""
```
you can use R"" as a raw string and get rid of \ inside of the string
that should be more readable

RomanDzhabarov commented:
I've done some initial pass over it

fabolive commented:
@RomanDzhabarov I've addressed your comments above. Please, se my responses to your comments.
Also, based on the issues you raised, I fixed things in other parts of the code (anticipating further comments from you).

mattklein123 commented:
@fabolive @RomanDzhabarov is on vacation, I will go through this today.

mattklein123 commented on source/common/tracing/zipkin/zipkin_core_types.h:
```diff
@@ -0,0 +1,586 @@
+ * Represents a Zipkin endpoint. This class is based on Zipkin's Thrift definition of an endpoint.
+ * Endpoints can be added to Zipkin annotations.
+ */
+class Endpoint : public ZipkinBase {
```
We already have a full set of interfaces for network addresses. Rework this to use Network::Address as a member. Then ipv4 and ipv6, etc. will be handled for you.

mattklein123 commented on source/common/tracing/zipkin/zipkin_core_types.h:
```diff
@@ -0,0 +1,586 @@
+  std::string value_;
+  Endpoint endpoint_;
+
+  bool isset_endpoint_;
```
For all of the isset_ stuff it would better if you just did Optional<Endpoint> etc. (Applies elsewhere)

mattklein123 commented on source/common/tracing/zipkin/zipkin_core_types.h:
```diff
@@ -0,0 +1,586 @@
+ * This struct identifies which of the optional attributes are set in a Span object.
+ * Each member is a one-bit boolean indicating whether or not the corresponding attribute is set.
+ */
+typedef struct SpanIsSet {
```
SpanAttributeSet, also removed typedef

mattklein123 commented on source/common/tracing/zipkin/zipkin_core_types.h:
```diff
@@ -0,0 +1,586 @@
+  int64_t timestamp_;
+  int64_t duration_;
+  uint64_t trace_id_high_;
+
```
remove newlines, make as many members const as possible.

htuch commented on source/common/tracing/zipkin/span_context.cc:
```diff
@@ -0,0 +1,122 @@
+
+const std::string& SpanContext::SPAN_CONTEXT_REGEX_STR() {
+  // ^([0-9,a-z]{16});([0-9,a-z]{16});([0-9,a-z]{16})((;(cs|sr|cr|ss))*)$
+  static std::string span_context_regex_str =
```
This also has dangerous destruction ordering behavior if it leaks across compilation units, see https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables. Prefer:

```
static const std::string* span_context_regex_str = new std::string(...);
return *span_context_regex_str;
```

htuch commented on source/common/tracing/zipkin/span_context.cc:
```diff
@@ -0,0 +1,122 @@
+}
+
+const std::regex& SpanContext::SPAN_CONTEXT_REGEX() {
+  static std::regex span_context_regex(SPAN_CONTEXT_REGEX_STR());
```
Same as above regarding static non-POD.

htuch commented on source/common/tracing/zipkin/span_context.cc:
```diff
@@ -0,0 +1,122 @@
+const std::string SpanContext::FIELD_SEPARATOR_ = ";";
+
+// String value corresponding to an empty span context.
+const std::string SpanContext::UNITIALIZED_SPAN_CONTEXT_ = "0000000000000000" + FIELD_SEPARATOR_ +
```
See other comments on constructor/destructor ordering across modules. Strongly prefer to allocate non-POD static/globals on heap via function local static pointers and leave them there.

mattklein123 replied:
Please fix all the POD stuff per @htuch comments.

htuch commented on source/common/tracing/zipkin/span_context.cc:
```diff
@@ -0,0 +1,122 @@
+
+  if (std::regex_search(span_context_str, match, SPAN_CONTEXT_REGEX())) {
+    // This is a valid string encoding of the context
+    trace_id_ = stoull(match.str(1), nullptr, 16);
```
std::stoull etc.

htuch commented on source/common/tracing/zipkin/util.cc:
```diff
@@ -0,0 +1,82 @@
+  std::regex re("^(\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3})(:(\\d+))?$");
+  std::smatch match;
+  if (std::regex_search(address, match, re)) {
+    ip = match.str(1);
```
I think there are better ways to get an IP/port out of an address that exist in the Envoy code base today. See https://github.com/lyft/envoy/blob/master/source/common/network/address_impl.cc#L223. This lredy hndles IPv6.

mattklein123 commented:
@fabolive if you merge master now, you can kill all the cmake files in this change.

fabolive commented:
@mattklein123 One more comment. The Zipkin library uses MonotonicTime to compute the span duration, but uses SystemTime for timestamps. Timestamps are rendered by the Zipkin UI as human-readable times.

mattklein123 commented:
@fabolive cool will look at this in the morning and let's see if we can get this merged tomorrow. I think we are pretty close.

mattklein123 commented:
Also, in the future, please don't rebase your change once you first push the PR. Just merge master and add commits. It makes it much easier to review the deltas.

mattklein123 commented on source/common/common/hex.cc:
```diff
@@ -43,3 +43,18 @@ std::vector<uint8_t> Hex::decode(const std::string& hex_string) {
 }
+
+std::string Hex::uint64ToHex(uint64_t value) {
+  std::vector<uint8_t> data(8);
```
Use std::array here to avoid heap allocation.

mattklein123 commented on source/common/tracing/zipkin/zipkin_core_types.h:
```diff
@@ -0,0 +1,536 @@
+  virtual const std::string& toJson() PURE;
+
+protected:
+  std::string json_string_;
```
It's pretty odd to have this be a member variable and basically use it as a temporary. I would get rid of this, and then have all the functions return a temporary. The compiler will optimize it.

fabolive commented:
@mattklein123 Please, see my latest commit to address your comments.
Note that, after I merged with master, check_format started failing for the files:
source/common/tracing/zipkin/BUILD
test/common/tracing/zipkin/BUILD
However, buildifier does not find anything wrong with the files. Any clue on how to fix this?

mattklein123 commented:
@fabolive I think you are missing license() directive. Just run envoy/tools/check_format.py fix. Anyway this looks good to me. @htuch wanted to take a look and we can merge on Monday.

htuch commented on test/common/tracing/zipkin/tracer_test.cc:
```diff
@@ -0,0 +1,214 @@
+  Network::Address::InstanceConstSharedPtr addr =
+      Network::Address::parseInternetAddressAndPort("127.0.0.1:9000");
+  Tracer tracer("my_service_name", addr);
+  tracer.setRandomGenerator(Runtime::RandomGeneratorPtr(new Runtime::RandomGeneratorImpl()));
```
Should use `MockRandomGenerator` for test determinism.

htuch commented on test/common/tracing/zipkin/tracer_test.cc:
```diff
@@ -0,0 +1,214 @@
+      Network::Address::parseInternetAddressAndPort("127.0.0.1:9000");
+  Tracer tracer("my_service_name", addr);
+  tracer.setRandomGenerator(Runtime::RandomGeneratorPtr(new Runtime::RandomGeneratorImpl()));
+  MonotonicTime start_time = ProdMonotonicTimeSource::instance_.currentTime();
```
Same here but with `MockMonotonicTimeSource`.

htuch commented on test/common/tracing/zipkin/tracer_test.cc:
```diff
@@ -0,0 +1,214 @@
+  Network::Address::InstanceConstSharedPtr addr =
+      Network::Address::parseInternetAddressAndPort("127.0.0.1:9000");
+  Tracer tracer("my_service_name", addr);
+  MonotonicTime start_time = ProdMonotonicTimeSource::instance_.currentTime();
```
Please mock with `MockMonotonicTimeSource` for test determinism.

htuch commented on source/common/tracing/zipkin/zipkin_json_field_names.h:
```diff
@@ -0,0 +1,31 @@
+  static const std::string ENDPOINT_SERVICE_NAME;
+  static const std::string ENDPOINT_PORT;
+  static const std::string ENDPOINT_IPV4;
+  static const std::string ENDPOINT_IPV6;
```
These are also all subject to the static initialization order fiasco. Read about this at:
* https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables
* https://isocpp.org/wiki/faq/ctors#static-init-order
* https://isocpp.org/wiki/faq/ctors#static-init-order-on-first-use
It might make sense to make these `const char[]` POD C strings. Or add some macros to reduce the boiler plate around initialization-on-first-use. 

I'm not bringing this up to be picky, but because this will fail ASAN when we switch on the check for this and it's really like playing with fire, since any small changes to the compiler or phase of the moon might lead to crashes during construction or destruction.

mattklein123 replied:
Another pattern to make changing this a little less painful is to actually make all of these non-static members, and have a static get() function to retrieve a singleton class. https://github.com/lyft/envoy/blob/master/source/common/http/headers.h#L102 (Though this example is not completely correct and needs to be fixed to allocate the instance on the heap.

htuch replied:
+1 on singleton. Let me see if I can factor this out into some base template class that folks can use to avoid having to simplify.

htuch replied:
https://github.com/lyft/envoy/pull/869

htuch commented on source/common/tracing/zipkin/zipkin_core_types.cc:
```diff
@@ -0,0 +1,270 @@
+  } else if ((context.isSetAnnotation().cs_) && (!context.isSetAnnotation().cr_)) {
+    // Need to set the CR annotation
+    Annotation cr;
+    uint64_t stop_timestamp =
```
Nit: const.

htuch commented on source/common/tracing/zipkin/zipkin_core_types.cc:
```diff
@@ -0,0 +1,270 @@
+    cr.setTimestamp(stop_timestamp);
+    cr.setValue(ZipkinCoreConstants::CLIENT_RECV);
+    annotations_.push_back(std::move(cr));
+    int64_t monotonic_stop_time =
```
Nit: const.

htuch commented on source/common/tracing/zipkin/zipkin_core_types.cc:
```diff
@@ -0,0 +1,270 @@
+void Span::finish() {
+  // Assumption: Span will have only one annotation when this method is called
+  SpanContext context(*this);
+  if ((context.isSetAnnotation().sr_) && (!context.isSetAnnotation().ss_)) {
```
Nit: unnecessary parentheses.

htuch commented on source/common/tracing/zipkin/zipkin_core_types.cc:
```diff
@@ -0,0 +1,270 @@
+  }
+
+  auto t = tracer();
+  if (t) {
```
Nit: `if (auto t = tracer()) {`

htuch commented on source/common/tracing/zipkin/zipkin_core_types.cc:
```diff
@@ -0,0 +1,270 @@
+}
+
+void Span::setTag(const std::string& name, const std::string& value) {
+  if ((name.size() > 0) && (value.size() > 0)) {
```
Nit: unnecessary parentheses.

htuch commented on source/common/tracing/zipkin/tracer.cc:
```diff
@@ -0,0 +1,122 @@
+
+    // Set the timestamp globally for the span
+    span_ptr->setTimestamp(timestamp_micro);
+  } else if ((previous_context.isSetAnnotation().cs_) &&
```
Nit: unnecessary parentheses (and all other places).

htuch commented on source/common/tracing/zipkin/util.h:
```diff
@@ -0,0 +1,47 @@
+  /**
+   * Returns a randomly-generated 64-bit integer number.
+   */
+  static uint64_t generateRandom64();
```
Why not always use `RandomGenerator()`? Is there a good reason to do random number generation differently here?

fabolive replied:
Originally, my intention was to write a Zipkin library independent from Envoy. The code on PR #700 was supposed to be the only glue between Envoy and the Zipkin library. That's why my original code (which is long gone after several code-review rounds) did not use any of the classes and utilities from the Envoy code base. That includes random-number generation.

fabolive replied:
I would like to keep this, at least, for now.

fabolive commented on source/common/common/hex.cc:
```diff
@@ -43,3 +44,18 @@ std::vector<uint8_t> Hex::decode(const std::string& hex_string) {
+  data[1] = (value & 0x00FF000000000000) >> 48;
+  data[0] = (value & 0xFF00000000000000) >> 56;
+
+  return encode(&data[0], data.size());
```
The reason is performance.

fabolive commented on source/common/tracing/zipkin/span_context.h:
```diff
@@ -0,0 +1,151 @@
+   * @param span_context_str The string-encoding of a SpanContext in the same format produced by the
+   * method serializeToString().
+   */
+  void populateFromString(const std::string& span_context_str);
```
This method is not for testing only. My original PR was split into two. This PR is only about a Zipkin "library"; the code that actually uses the "library" to perform tracing is part of PR #700 (which is outdated now).  The method populateFromString is needed by the other PR.

fabolive commented on source/common/tracing/zipkin/span_context.h:
```diff
@@ -0,0 +1,151 @@
+  uint64_t id_;
+  uint64_t parent_id_;
+  AnnotationSet annotation_values_;
+  bool is_initialized_;
```
As per my comment above, populateFromString is not used for tests only.

fabolive commented on source/common/tracing/zipkin/span_buffer.cc:
```diff
@@ -0,0 +1,30 @@
+    for (uint64_t i = 1; i < size; i++) {
+      stringified_json_array += ",";
+      stringified_json_array += span_buffer_[i]->toJson();
+    }
```
Well, Json::Factory::listAsJsonString() does not do exactly the same thing. As a result of that method, the elements of the stringified json array are stringified as well, which means that an extra-level of double-quote escaping would be present. That was not my intention.
I would, therefore, prefer to keep this code unchanged.

fabolive commented on source/common/tracing/zipkin/span_buffer.h:
```diff
@@ -0,0 +1,63 @@
+   * Constructor that creates an empty buffer. Space needs to be allocated by invoking
+   * the method allocateBuffer(size).
+   */
+  SpanBuffer() {}
```
Yes, this is used by the code on PR #700. PR #700, which is supposed to be merged after this, makes use of the classes on this PR to perform the actual Zipkin tracing.

fabolive commented on source/common/tracing/zipkin/span_buffer.h:
```diff
@@ -0,0 +1,63 @@
+   *
+   * @param size The desired buffer size.
+   */
+  SpanBuffer(uint64_t size) { span_buffer_.reserve(size); }
```
Fair enough.

fabolive commented on source/common/tracing/zipkin/span_context.cc:
```diff
@@ -0,0 +1,131 @@
+SpanContext::SpanContext(const Span& span) {
+  trace_id_ = span.traceId();
+  id_ = span.id();
+  parent_id_ = span.isSetParentId() ? span.parentId() : 0;
```
These fields are not supposed to be `const`. The method populateFromString() changes the object state, i.e., it changes the values of those fields.

fabolive commented on source/common/tracing/zipkin/span_context.cc:
```diff
@@ -0,0 +1,131 @@
+          annotation_values_.sr_ = true;
+        } else if (annotation_value == ZipkinCoreConstants::SERVER_SEND) {
+          annotation_values_.ss_ = true;
+        }
```
A SpanContext can be created from either a Span or from a string. Those are two distinct uses of this class. In fact, the ability to create a Span from a string is critical to the code of PR #700.

fabolive commented:
@htuch Please, see my commits above, as well as my replies to your comments.
cc: @mattklein123, @rshriram

htuch commented:
I'll make another pass this arvo.

fabolive commented:
@htuch I will make one more commit. I have applied the singleton constant template to zipkin_json_field_names.h, but forgot to apply it to zipkin_core_constants.h. I will commit shortly.

fabolive commented:
@htuch, @mattklein123. I have just pushed one more commit. This should be the last planned one.

htuch commented on source/common/tracing/zipkin/span_context.cc:
```diff
@@ -0,0 +1,152 @@
+          annotation_values_.sr_ = true;
+        } else if (annotation_value == ZipkinCoreConstants::get().SERVER_SEND) {
+          annotation_values_.ss_ = true;
+        }
```
There some commonality in this if-else block with the constructor block that does the same. Please consider refactoring in the future.

fabolive replied:
@htuch, @mattklein123, @RomanDzhabarov, @adriancole : Thanks for your reviews!
@htuch: I will make sure the other PRs are smaller and will pull you early on in the process. I will consider the code refactoring suggested above.
I will resume the work on PR #700 and let you guys know when you can take a look at it. The code in that branch is outdated now.

adriancole commented:
Hey folks Fyi #1196 aims to remove this work, pushing it to a separate repository making changes to support zipkin indirect. I'm concerned as it seems to be discussed without many of you providing input. I just want to make sure you are cool or voices are heard otherwise. If you are cool, cool with me.
